src/manifolds/Symplectic.jl

```diff
@@ -0,0 +1,889 @@
+function (Base.:+)(p::AbstractMatrix, Q::SymplecticMatrix)
+    # When we are adding, the Matrices must match in size:
+    n, _ = get_half_dims(p, true, false, true)
+
+    # Allocate new memory:
+    TS = Base._return_type(+, Tuple{eltype(p),eltype(Q)})
+    out = copyto!(similar(p, TS), p)
+
+    # Add Q.λ multiples of the UniformScaling to the lower left and upper right blocks of p:
+    λ_Id = LinearAlgebra.UniformScaling(Q.λ)
+
+    out[1:n, (n + 1):(2n)] += λ_Id
+    out[(n + 1):(2n), 1:n] -= λ_Id
+    return out
+end
+
+# Binary minus:
+(Base.:-)(Q::SymplecticMatrix, p::AbstractMatrix) = Q + (-p)
+(Base.:-)(p::AbstractMatrix, Q::SymplecticMatrix) = p + (-Q)
+
+function (Base.:*)(Q::SymplecticMatrix, p::AbstractVecOrMat)
+    n, _ = get_half_dims(p, true, false)
+
+    # Allocate new memory:
+    TS = typeof(one(eltype(p)) + one(eltype(Q)))
+    Qp = similar(p, TS)
+
+    # Perform left mulitply by λ*Q:
+    mul!((@inbounds view(Qp, 1:n, :)), Q.λ, @inbounds view(p, (n + 1):lastindex(p, 1), :))
+    mul!((@inbounds view(Qp, (n + 1):lastindex(Qp, 1), :)), -Q.λ, @inbounds view(p, 1:n, :))
+
+    return Qp
+end
+
+function (Base.:*)(p::AbstractVecOrMat, Q::SymplecticMatrix)
+    _, k = get_half_dims(p, false, true)
+
+    # Allocate new memory:
+    TS = typeof(one(eltype(p)) + one(eltype(Q)))
+    pQ = similar(p, TS)
+
+    # Perform right mulitply by λ*Q:
+    mul!((@inbounds view(pQ, :, 1:k)), -Q.λ, @inbounds view(p, :, (k + 1):lastindex(p, 2)))
+    mul!((@inbounds view(pQ, :, (k + 1):lastindex(pQ, 2))), Q.λ, @inbounds view(p, :, 1:k))
+    return pQ
+end
+
+function LinearAlgebra.lmul!(Q::SymplecticMatrix, p::AbstractVecOrMat)
+    # Perform left multiplication by a symplectic matrix,
+    # overwriting the matrix p in place:
+    n, _ = get_half_dims(p, true, false)
+
+    half_row_p = copy(@inbounds view(p, 1:n, :))
+
+    mul!((@inbounds view(p, 1:n, :)), Q.λ, @inbounds view(p, (n + 1):lastindex(p, 1), :))
+
+    mul!(
+        (@inbounds view(p, (n + 1):lastindex(p, 1), :)),
+        -Q.λ,
+        @inbounds view(half_row_p, :, :)
+    )
+    return p
+end
+
+function LinearAlgebra.rmul!(p::AbstractVecOrMat, Q::SymplecticMatrix)
+    # Perform right multiplication by a symplectic matrix,
+    # overwriting the matrix p in place:
+    _, k = get_half_dims(p, false, true)
+
+    half_col_p = copy(@inbounds view(p, :, 1:k))
+
+    mul!((@inbounds view(p, :, 1:k)), -Q.λ, @inbounds view(p, :, (k + 1):lastindex(p, 2)))
+
+    mul!(
+        (@inbounds view(p, :, (k + 1):lastindex(p, 2))),
+        Q.λ,
+        @inbounds view(half_col_p, :, :)
+    )
+
+    return p
+end
+
+function LinearAlgebra.mul!(A::AbstractVecOrMat, Q::SymplecticMatrix, p::AbstractVecOrMat)
+    n, _ = get_half_dims(p, true, false)
+    # Perform left multiply by λ*Q:
+    mul!((@inbounds view(A, 1:n, :)), Q.λ, @inbounds view(p, (n + 1):lastindex(p, 1), :))
+    mul!((@inbounds view(A, (n + 1):lastindex(A, 1), :)), -Q.λ, @inbounds view(p, 1:n, :))
```

Comment:
If you `@inbounds` access to `A`, you need to check its size first like here: https://github.com/JuliaLang/julia/blob/b486f0dec2f0049a22d263cc0e39f374a1c3b40a/stdlib/LinearAlgebra/src/symmetric.jl#L205-L214 for example.

Reply:
Good catch, I've added checks for the left- and right-multiply functions for `SymplecticMatrix`.